Add Automated updating of `autocomplete-html` `completions.json`

confused-Techie opened the pull request:
Alright this is a followup to #398 where the whole update process for `autocomplete-css` was rewritten.

Again this turned out to be more involved than originally thought. Since the previous update method points to broken URLs this had to be rewritten entirely.

With this PR I've gone ahead and added a new update method and placed all files needed to update within a folder titled `update`. And due to path concerns I've added a script to the `package.json` called `update` that can be run to create the updated files as needed. This is noted in the `README.md` as the way to update the `completions.json`.

That's again the short version, but for some context:

---

Previously `autocomplete-html` would use `https://raw.githubusercontent.com/adobe/brackets/master/src/extensions/default/HTMLCodeHints/HtmlTags.json` and `https://raw.githubusercontent.com/adobe/brackets/master/src/extensions/default/HTMLCodeHints/HtmlAttributes.json` to generate their completions. But as the project that created those files [`adobe/brackets`](https://github.com/adobe/brackets/) is archived and no longer maintained, these files could no longer serve as a way to update the `completions.json` file.

The other difficult aspect of keeping this updated is there doesn't seem to be any reliable source of data that we could obtain that has an easily machine readable HTML spec to refer to. Nearly every solution would require quite a bit of finagling to get working. So instead I've opted to rely on some good sources of data, that I hope we can continue to do so for some time.

* We rely on MDN, and the [Markdown documents](https://github.com/mdn/content/) that make up their website to obtain Element Descriptions.
* We rely on [`w3c/webref`](https://github.com/w3c/webref/) to obtain every Element that exists within the standard
* And lastly we rely on [Chromium's DevTools](https://github.com/ChromeDevTools/devtools-frontend/) to obtain all valid attributes of each element.

The last item on this list is likely the most controversial. Essentially from my research there was no single source of valid attributes for elements besides the [`whatwg/html`](https://raw.githubusercontent.com/whatwg/html/main/source) full HTML spec which is written in a `proprietary language that is then post-processed into HTML`. So rather than build a full parser of this data I opted to rely on Chromium's DevTools. The reason for this, is because they too rely on the same Element data from `w3c/webref` and themselves keep a manually maintained list of valid attributes. 

So while there is a bit of hackiness (see `chromium-elements-shim.js` and `chromium-elements-shim.mjs`) to be able to use one of their repositories files directly, it does provide us with a super simplistic way to obtain literally every valid attribute for nearly every element defined in any specification recognized by W3C.

A point of concern:

Like I just said above, our `tags` element of the attributes contains `literally every valid attribute for nearly every element defined in any specification recognized by W3C`, which even further our `tags` element, still contains literally every valid element defined in any specification recognized by W3C. The above note is just if we are able to grab attributes for them.

What this does mean, is our `completions.json` file has increased drastically. As in the size of our `completions.json` file (measured in lines) has increased `~1,301.51%`. From originally `1,919` lines to `26,895` lines.

So while already some users on Discord have suggested this might be to much, I wanted to put up a PR so people could actually read through the code and new `completions.json` to help form an opinion on what should be done.

---

What to do of the file size increase:

So like said before some users on Discord have suggested we should exclude items that appear in the final `completions.json` file.

And there's a few options we can look at for this, but first I do want to provide my opinion.
I feel like we should be in no position to limit or curate anything that does exist in the actual spec for our completions. The completions provided should be able to help out the developer working on the cutting edge, and the developer having to maintain a 30 year old website, using outdated and non-standard elements. That is as long as something is officially in a standard provided by W3C, we should help them with autocompletions.

The flip side to this, which are realistic concerns, is this file size increase will increase the size of the Pulsar binary. Which for reference now, after updating, the `completions.json` is `628 KB`. While currently, without my PR, the `completions.json` file is `48.0 KB`. Additionally loading such a large file into memory of a default package of Pulsar could lead to a slight increase in RAM usage on machines. I don't think this is the biggest concern though honestly.
Then lastly it could present some users to see non-standard items during autocompletions. Now autocompletions are already provided based on what you're typing, and will be ordered to show what's closer to your current typings as always. But that was the concern provided on Discord, so I'll mention it here.

Now for this we have some options. 

1) We do nothing, we let the increase just be as it is, more valid autocompletions for our users.
2) We manually curate a blacklist of items to exclude. This option does mean our automated process is no longer fully automated. But could reduce completions by a substantial amount.
3) We make it configurable. A user has a setting that can specify whether to show a `curated` autocompletions, or `full`. Or could even allow autocompletions to be turned on or off by the specification. Such as `html-spec`, `svg11-spec`, `mathml-core-spec` and so on, for all specs we provide.

No obviously, from my opinion above I'm against removing things wholesale, and if people did want to have a smaller autocomplete I'd be much more inclined to provide a configurable experience for our users. But since when sharing the new data about the `completions.json` the first and only response of other community members was that we needed to remove items I did want to mention this and let more opinions on the matter in.

---

But otherwise, I know another long writeup and huge diff, but please feel free to ask any questions as needed, and hope we can see this get merged!

Resolves #393 

PS. A quick note, I have not yet touched tests. They will fail. I first wanted to see what thoughts people had, and complete the package if we needed to make changes before having to fix tests, since much like `autocomplete-css` they seem to completely rely on the order and amount of autocompletions that would be expected of different typings, and will need substantial changes to get passing again.

mauricioszabo commented on packages/autocomplete-html/update/update.js:
```diff
@@ -0,0 +1,289 @@
+  it into our completions during update time.
+ */
+
+const chromiumElementsShim = require("./chromium-elements-shim.js");
```
We can probably get rid of this `.js` that delegates to `.mjs` if we use `import` in place of `require`.

`import` will happily accept a ESM inside a CommonJS, and the result will be a promise, so we just need to `await` for it to resolve... like so:

```suggestion
const chromiumElementsShim = import("./chromium-elements-shim.mjs");
```

confused-Techie replied:
I'll try this. Although I think I did something similar and it was failing. But that was likely before I had implemented the translation from TypeScript

confused-Techie replied:
I'll try this. Although I think I did something similar and it was failing. But that was likely before I had implemented the translation from TypeScript

confused-Techie replied:
As for this one, doing some more research after your suggestion was failing "CommonJS modules cannot import ES Modules"

So it seems we might have to keep this hacky way to keep things going. Plus because the package we use to convert TypeScript on the fly is also an ESM module. So unfortunately we might not be able to fix this right away.

mauricioszabo commented:
Just for the record: I made a suggestion about using a `import` instead of wrapping up around a require, but I'm not 100% sure it works - so, maybe test it :)

mauricioszabo commented on packages/autocomplete-html/update/update.js:
```diff
@@ -0,0 +1,289 @@
+  When an entry contains no `attributes` there is no empty array, the element
+  simply doesn't exist.
+
+  The `attributes` object contains keys of different elements that themselves
+  are objects that can contain several valid keys.
+  - global: Seems to be used exclusively for Global Attributes. Is a boolean
+            which when false, the key does not appear.
+  - type: A ?type? for the attribute. It's meaning is not immediately known.
+          Nor a way to reliabley programatically collect it. Some discovered values:
+          * cssStyle: Exclusively used for `class` attribute
+          * boolean: Attributes that only accept `true` or `false`
+          * flag: For attributes that don't require or accept values. eg autoplay
+          * cssId: Exclusively used for the `id` attribute
+          * color: Exclusively used for the `bgcolor` attribute
+          * style: Exclusively used for the `style` attribute
+  - description: A text description of the attribute
+  - attribOption: A string array of valid values that can exist within the attribute.
+                  Such as the case with `rel` where only so many valid options exist.
+  ==============================================================================
+  ==============================================================================
+  A quick check in for the structure of `@webref/elements` `listAll()` return.
+  This will return an object, where each top level key is a HTML spec name.
+  Then within each will be some information on the spec, and `elements` an array
+  of objects that will contain the `name` of our element, and it's interface name.
+  {
+    "css-masking-1": {
+      "spec": {},
+      "elements": [
+        { "name": "clipPath", "interface": "SVGClipPathElement" },
+        { "name": "mask", "interface": "SVGMaskElement" }
+      ]
+    }
+  };
+
+  The `interface` here should match a return that we can receive from DevTools data.
+  ==============================================================================
+  ==============================================================================
```
I honestly would prefer these to become a test instead of a comment...

confused-Techie replied:
Yeah I did write a whole lot here.

But I can keep some comments to explain what can't be done by a test maybe? I'll look at adding Joi to this so it fails early if we get an unexpected format. But thanks for the suggestion!

mauricioszabo commented on packages/autocomplete-html/update/update.js:
```diff
@@ -0,0 +1,289 @@
+    element = substitueElements[element];
+  }
+
+  let file;
+
+  // First lets find the file, but when not initially found, we will continue
+  // to search valid locations. Since MDN has content of valid tags seperated
+  // by essentially the spec they exist in.
+  if (fs.existsSync(`./node_modules/content/files/en-us/web/html/element/${element}/index.md`)) {
+    file = fs.readFileSync(`./node_modules/content/files/en-us/web/html/element/${element}/index.md`, { encoding: "utf8" });
+  } else if (fs.existsSync(`./node_modules/content/files/en-us/web/svg/element/${element}/index.md`)) {
+    file = fs.readFileSync(`./node_modules/content/files/en-us/web/svg/element/${element}/index.md`, { encoding: "utf8" });
+  } else if (fs.existsSync(`./node_modules/content/files/en-us/web/mathml/element/${element}/index.md`)) {
+    file = fs.readFileSync(`./node_modules/content/files/en-us/web/mathml/element/${element}/index.md`, { encoding: "utf8" });
+  }
+
+  if (typeof file === "string") {
```
```suggestion
  // First lets find the file, but when not initially found, we will continue
  // to search valid locations. Since MDN has content of valid tags seperated
  // by essentially the spec they exist in.
  const filePath = ['html', 'svg', 'mathml'].map(path =>
    `./node_modules/content/files/en-us/${path}/html/element/${element}/index.md`
  ).find(f => fs.existsSync(f));

  if (filePath) {
    file = fs.readFileSync(filePath, {encoding: "utf8"});
```

mauricioszabo replied:
WDYT?

confused-Techie replied:
Oh yeah this is sick! Only difference being we would want the path to look like `./node_modules/content/files/en-us/web/${path}/element/${element}/index.md`

But I'll make that change her in a second

confused-Techie replied:
Added this one, just with a fix to the path

mauricioszabo commented on packages/autocomplete-html/update/update.js:
```diff
@@ -0,0 +1,289 @@
+    // In case the first few lines is an empty line break
+    for (let i = 0; i < summaryRaw.length; i++) {
+      if (summaryRaw[i].length > 1) {
+        return summaryRaw[i]
+                .replace(/\{\{\S+\("(\S+)"\)\}\}/g, '$1')
+                .replace(/\*/g, "")
+                .replace(/\`/g, "")
+                .replace(/\{/g, "")
+                .replace(/\}/g, "")
+                .replace(/\"/g, "")
+                .replace(/\[([A-Za-z0-9-_* ]+)\]\(\S+\)/g, '$1');
```
This can work:

```suggestion
        return summaryRaw[i]
                .replace(/\{\{\S+\("(\S+)"\)\}\}/g, '$1')
                .replace(/[\*\`\{\}\"]/g, "")
                .replace(/\[([A-Za-z0-9-_* ]+)\]\(\S+\)/g, '$1');
```

But I feel it's a good idea to explain what exactly this regexp is trying to replace -again, possibly with a test because it will future-proof refactorings...

confused-Techie replied:
Yeah this part is explained a bit better I think in the new PR for `autocomplete-css` as it's exactly copied from there. But you do have a point

confused-Techie replied:
Added your simplification, as well as some more comments to help explain what the intention is

mauricioszabo commented on packages/autocomplete-html/update/update.js:
```diff
@@ -0,0 +1,289 @@
+  // Some elements are similar enough they exist in a single folder of a different
+  // name. So we will manually intervein in those cases.
+
+  const substitueElements = {
+    "h1": "heading_elements",
+    "h2": "heading_elements",
+    "h3": "heading_elements",
+    "h4": "heading_elements",
+    "h5": "heading_elements",
+    "h6": "heading_elements"
+  };
+
+  if (substitueElements[element]) {
+    element = substitueElements[element];
+  }
```
I feel we can make this easier by just using a regexp:

```suggestion
  if (element.match(/^h[1-6]$/)) {
    element = 'heading_elements';
  }
```

confused-Techie replied:
Yeah good point. I originally wrote it not knowing how many elements would need substitutions, but this looks good to me, thanks!

confused-Techie commented:
So @mauricioszabo I've gone ahead and tried to address your concerns.

I've added built in validation to the object structures that we expect via `joi`.  This allows us to test and properly document object structure in one place without the need of including it in the huge comment at the top of the file.

Additionally I've moved the heaviest piece of Regex to it's own function and added some tests with `jest` (which I know isn't your favorite, and I can try and implement this in Jasmine if we prefer, it's just what I know best, and for a one off like this made sense for me. This small bit of testing is not run as the package test, since it doesn't make to much sense to me to add it there. Rather it is a dedicated script that can be called with `npm run update:test` and will just run a few quick tests to make sure the Regex is working as expected against static strings.

If we do feel additional testing is required it can be done, but would require quite a bit of refactoring, since nearly every function tries interacting with the file system in some way. (Or at least the ones we would want to test)

And lastly, now knowing the final implementation I went through the original block comment and pruned it down only to what I felt was necessary for future devs looking at this code to know what was going on.

---

Otherwise though all other suggestions have been made, except the hacky way of requiring the DevTools source code, as I couldn't find any other way to do it. Please let me know what you think, and again appreciate you taking a look at this PR!

confused-Techie commented:
Alright, so some updates.

I know it's been a while since anything was done on this PR, but I've gone ahead and made some updates, that would've had to be made anyways once it started going under review.

### Tests 

The tests here much like `autocomplete-css` did a whole lot to essentially test the exact order of completions, and their exact values. Since we obviously can no longer ensure any explicit order that's been removed. We now check that if certain tags are anywhere in the completions file. That's what the additional functions in the spec do, enumerate all values and determine if what we want is included. While these functions also contain optional values to check specific keys on the completions object for a value there.

Additionally many tests would check the exact length of items within completions, this has been changed to instead check that the amount of completions is over whatever limit they originally checked. The full justification and conversation about that is over on #398. But the short version is that @DeeDeeG and I thought it was more likely for an item to be added to the spec that matches the completion rather than removed. So by just ensuring it's higher than whatever number it was previously set as means we have to think about whatever that number is much less.

Lastly some tests would test for certain descriptions or URL's on specific values. At first this was swapped out by instead ensuring the description existed at all, but for some values I've also added a function to return the index of a specific value, which we can then check it's specific description or other items.

### Updating the Lib 

While I intended to stay away from changing any of the actual package's code, I did notice that the URL's being generated for attributes of items would no longer work. Originally this would craft the URL like `URL#attr-<attribute-name>` but it seems now that on MDN's website individual attributes don't get a specific anchor, we instead should like like `URL#attributes`. So I've fixed that in the packages code while updating the test that checks for this.

### Globals 

After a lot of checking within the tests, and I was able to confirm that our `curated-attributes` is supposed to only contain global attributes/properties. Which was true, but it also meant that individual completions otherwise should NOT contain any global values, which they did.

So what I've done is updating the `update.js` script to never inject a global event or attribute into our end result completions. 

But because our `curated-attributes` document is static, we don't have any way to update newer globals, so I've now added a check at the end of the update script that outputs any global values found while updating, if they weren't also found in our existing `curated-attributes` file, so in that way that document can be much more easily kept up to date.

---

So I'll work to get our `curated-attributes` document manually updated with 32 missing global attributes, but any review on this existing code would be highly appreciated.

confused-Techie commented:
Alright this PR is now ready to be merged if found suitable. I've updated our static `curated-attributes` to include all currently supported global values. As well as all tests should be (and as of the last run) are passing. 

So please feel free to give this one a look, and we can be done with all of this autocompletion stuff, up until the next time we decide to rerun the update scripts

confused-Techie commented:
So on this PR:

* `find-and-replace`: 54 Failures ✅
* `settings-view`: 1 Failure ✅ (Up till recently we expected 0, but we are seeing 1 across the board. So until that is resolved)
* `symbols-view`: 2 Failures ✅
* `tree-view`: 2 Failures ✅

So this PR is ready to review, with all tests fully passing.

confused-Techie commented:
As discussed on Discord, gonna go ahead and merge this one. Thanks for all those that provided help getting it fleshed out properly!
